fvm/evm/emulator/emulator.go

```diff
@@ -0,0 +1,318 @@
+	// would stay in memory for faster transaction execution. you
+	// have to explicitly ask the trie to commit to the underlying storage
+	err = proc.state.Database().TrieDB().Commit(newRoot, false)
+	if err != nil {
+		return gethTypes.EmptyRootHash, handleCommitError(err)
+	}
+	return newRoot, nil
+}
+
+func handleCommitError(err error) error {
+	if err == nil {
+		return nil
+	}
+	// if known types (database errors) don't do anything and return
+	if types.IsAFatalError(err) || types.IsADatabaseError(err) {
+		return err
+	}
+
+	// else is a new fatal error
+	return types.NewFatalError(err)
+}
+
+func (proc *procedure) mintTo(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+		TxType:      types.DirectCallTxType,
+	}
+
+	// create account if not exist
+	if !proc.state.Exist(addr) {
+		proc.state.CreateAccount(addr)
+	}
+
+	// add balance
+	proc.state.AddBalance(addr, amount)
+
+	// we don't need to increment any nonce, given the origin doesn't exist
+	res.StateRootHash, err = proc.commit()
+
+	return res, err
+}
+
+func (proc *procedure) withdrawFrom(address types.Address, amount *big.Int) (*types.Result, error) {
+	var err error
+
+	addr := address.ToCommon()
+	res := &types.Result{
+		GasConsumed: proc.config.DirectCallBaseGasUsage,
+		TxType:      types.DirectCallTxType,
+	}
+
+	// check if account exists
+	// while this method is only called from bridged accounts
+	// it might be the case that someone creates a bridged account
+	// and never transfer tokens to and call for withdraw
+	if !proc.state.Exist(addr) {
+		return res, types.ErrAccountDoesNotExist
+	}
+
+	// check the source account balance
+	// if balance is lower than amount needed for withdrawal, error out
+	if proc.state.GetBalance(addr).Cmp(amount) < 0 {
+		return res, types.ErrInsufficientBalance
+	}
+
+	// sub balance
+	proc.state.SubBalance(addr, amount)
+
+	// we increment the nonce for source account cause
+	// withdraw counts as a transaction
+	nonce := proc.state.GetNonce(addr)
+	proc.state.SetNonce(addr, nonce+1)
+
+	res.StateRootHash, err = proc.commit()
+	return res, err
+}
+
+func (proc *procedure) run(msg *gethCore.Message, txType uint8) (*types.Result, error) {
+	res := types.Result{
+		TxType: txType,
+	}
+
+	gasPool := (*gethCore.GasPool)(&proc.config.BlockContext.GasLimit)
+	execResult, err := gethCore.NewStateTransition(
+		proc.evm,
+		msg,
+		gasPool,
+	).TransitionDb()
+	if err != nil {
+		res.Failed = true
+		// if the error is a fatal error or a non-fatal database error return it
+		if types.IsAFatalError(err) || types.IsADatabaseError(err) {
+			return &res, err
+		}
+		// otherwise is a validation error (pre-check failure)
+		// no state change, wrap the error and return
+		return &res, types.NewEVMValidationError(err)
+	}
+
+	// if prechecks are passed, the exec result won't be nil
+	if execResult != nil {
```

Comment:
do we even need to have this check? if err is nil this shouldn't be. I guess it doesn't hurt if the implementation details change, but I'm not sure there is a case in te future for both values to be nil.

Reply:
we can't rely on the way evm return results and error, in some cases it might be nil and error also be nil. better be safe than crashing.